localstack-core/localstack/services/lambda_/event_source_mapping/pollers/kinesis_poller.py

```diff
@@ -135,3 +142,57 @@ def get_sequence_number(self, record: dict) -> str:
             return record["kinesis"]["sequenceNumber"]
         else:
             return record["sequenceNumber"]
+
+    def pre_filter(self, events: list[dict]) -> list[dict]:
+        # TODO: test what happens with a mixture of data and non-data filters?
+        if has_data_filter_criteria_parsed(self.filter_patterns):
+            parsed_events = []
+            for event in events:
+                raw_data = self.get_data(event)
+                try:
+                    data = self.parse_data(raw_data)
+                    # TODO: test "data" key remapping
+                    # Filtering remaps "kinesis.data" in ESM to "data (idempotent for Pipes using "data" directly)
+                    # ESM: https://docs.aws.amazon.com/lambda/latest/dg/with-kinesis-filtering.html
+                    # Pipes: https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-kinesis.html
+                    # Pipes filtering: https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-event-filtering.html
+                    parsed_event = deepcopy(event)
+                    parsed_event["data"] = data
+
+                    parsed_events.append(parsed_event)
+                except json.JSONDecodeError:
+                    LOG.warning(
+                        f"Unable to convert event data '{raw_data}' to json... Record will be dropped.",
+                        exc_info=LOG.isEnabledFor(logging.DEBUG),
+                    )
+            return parsed_events
+        else:
+            return events
+
+    def post_filter(self, events: list[dict]) -> list[dict]:
+        if has_data_filter_criteria(self.filter_patterns):
+            # convert them back (HACK for fixing parity with v1 and getting regression tests passing)
+            for event in events:
+                parsed_data = event.pop("data")
+                encoded_data = self.encode_data(parsed_data)
+                self.set_data(event, encoded_data)
+        return events
+
+    def get_data(self, event: dict) -> str:
+        if self.kinesis_namespace:
+            return event["kinesis"]["data"]
+        else:
+            return event["data"]
+
+    def set_data(self, event: dict, data: bytes) -> None:
+        if self.kinesis_namespace:
+            event["kinesis"]["data"] = data
+        else:
+            event["data"] = data
+
+    def parse_data(self, raw_data: str | dict) -> dict | str:
+        decoded_data = base64.b64decode(raw_data)
+        return json.loads(decoded_data)
+
+    def encode_data(self, parsed_data: dict) -> dict | str | bytes:
```

Comment:
How could that method return anything but "bytes"?

Reply:
fixed to `str`
(leftover from polymorphic version before I knew that's only a Kinesis thing)